### ISSUE TITLE
LInter 'typecheck' reports false positive imports not matching module name

### ISSUE THREAD

**Issue by e-chip** [golangci/golangci-lint#2783]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc).
- [X] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

### Description of the problem

`typecheck` linter reports false positive errors on imports for packages which have name different from the module name.

For example `github.com/oschwald/maxminddb-golang` module's package is named `maxminddb`

It is legal to import package as is and reference it as `maxminddb` but `typecheck` produces 2 errors:
1. Unused import
2. Undeclared name

As a workaround import alias can be added.

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.45.2 built from 8bdc4d3f on 2022-03-24T11:51:26Z
```

</details>

### Configuration file

any (running with `--no-config` option)

### Go environment

<details>

```console
$ go version && go env
go version go1.18.1 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/root/.cache/go-build"
GOENV="/root/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.18.1"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/project/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2621028960=/tmp/go-build -gno-record-gcc-switches"

```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run  -v --no-config --disable-all -E typecheck
INFO [lintersdb] Active 1 linters: [typecheck]    
INFO [loader] Go packages loading at mode 575 (types_sizes|files|imports|name|compiled_files|deps|exports_file) took 1.2010904s 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 1.6396ms 
INFO [linters context/goanalysis] analyzers took 12µs with top 10 stages: typecheck: 12µs 
INFO [runner] Processors filtering stat (out/in): skip_dirs: 2/2, exclude-rules: 2/2, diff: 2/2, filename_unadjuster: 2/2, path_prettifier: 2/2, exclude: 2/2, source_code: 2/2, path_shortener: 2/2, sort_results: 2/2, cgo: 2/2, uniq_
by_line: 2/2, max_from_linter: 2/2, severity-rules: 2/2, nolint: 2/2, autogenerated_exclude: 2/2, identifier_marker: 2/2, max_per_file_from_linter: 2/2, max_same_issues: 2/2, path_prefixer: 2/2, skip_files: 2/2 
INFO [runner] processing took 4.408ms with stages: nolint: 1.5305ms, source_code: 1.3128ms, path_prettifier: 1.2701ms, skip_dirs: 190.7µs, identifier_marker: 52.1µs, exclude-rules: 39.1µs, max_same_issues: 2.7µs, uniq_by_line: 2.6µs
, path_shortener: 1.7µs, cgo: 1.1µs, max_from_linter: 900ns, max_per_file_from_linter: 800ns, autogenerated_exclude: 800ns, filename_unadjuster: 700ns, sort_results: 400ns, exclude: 300ns, skip_files: 300ns, diff: 200ns, severity-ru
les: 100ns, path_prefixer: 100ns
INFO [runner] linters took 93.1299ms with stages: typecheck: 88.6259ms
main.go:6:13: undeclared name: `maxminddb` (typecheck)
        db, err := maxminddb.Open("data.mmdb")
                   ^
main.go:3:8: "github.com/oschwald/maxminddb-golang" imported but not used (typecheck)
import "github.com/oschwald/maxminddb-golang"
       ^
INFO File cache stats: 1 entries of total size 166B
INFO Memory: 15 samples, avg is 48.6MB, max is 52.5MB
INFO Execution took 1.3034543s

```

</details>

### Code example or link to a public repository

<details>

```go
package main

import "github.com/oschwald/maxminddb-golang"

func main() {
	db, err := maxminddb.Open("data.mmdb")
	if err != nil {
		panic(err)
	}
	_ = db.Close()
}
```

</details>

**Comment by ldez** [comment]
Hello,

`typecheck` is not a real linter it's just a way to parse/display "compilation" and linters errors (linter reports are not errors)

**Comment by e-chip** [comment]
Thanks for answering!
But still this is a false positive linter error.

Fun fact: I have tried to run on Windows and linter does not report an issue. But on Linux it does.

`go mod tidy` does not fix the problem as it is not related to the problem.
Reason is that `module` name does not match `package` name.
So in maxminddb package in `go.mod` we have 

```go
module github.com/oschwald/maxminddb-golang
```

But in `*.go`  files we have

```go
package maxminddb
```

I undestand that `typecheck` is not a linter but a way to catch compilation errors. But is is definitely not an error.

Probably this issue should be addressed to `golang.org/x/tools/go/analysis` maintainers team...

**Comment by ldez** [comment]
As you can see in my previous comment I don't have any issue with your code example (I'm on Linux)

**Comment by e-chip** [comment]
That's weird. Will try to investigate further. Thanks for your time!

**Comment by obacarol** [comment]
Same problem here. :crying_cat_face: 
Any suggestion?

**Comment by e-chip** [comment]
> Same problem here. 😿 Any suggestion?

Maintainers say there's no problem :)

For me import alias has solved the problem.

**Comment by sten4eg** [comment]
I have this problem too
